inline.h

```diff
@@ -842,53 +842,69 @@ Perl_rpp_replace_2_1(pTHX_ SV *sv)
 }
 
 
+/* Private helper function for _NN and _IMM_NN variants.
+ * Assumes sv has already had its ref count incremented,
+ * ready for being put on the stack.
+ * Intended to be small and fast, since it's inlined into many hot parts of
+ * code.
+ */
+
 PERL_STATIC_INLINE void
-Perl_rpp_replace_2_1_NN(pTHX_ SV *sv)
+Perl_rpp_replace_2_1_COMMON(pTHX_ SV *sv)
 {
-    PERL_ARGS_ASSERT_RPP_REPLACE_2_1_NN;
 
     assert(sv);
-    assert(PL_stack_sp[0]);
-    assert(PL_stack_sp[-1]);
 #ifdef PERL_RC_STACK
+    SV *sv2 = *PL_stack_sp--;
+    assert(sv2);
+    SV *sv1 = *PL_stack_sp;
+    assert(sv1);
+
+    *PL_stack_sp = sv;
     assert(rpp_stack_is_rc());
-    /* replace PL_stack_sp[-1] first; leave PL_stack_sp[0] in place while
-     * we free [-1], so if an exception occurs, [0] will still be freed.
-     */
-    SV *oldsv = PL_stack_sp[-1];
-    PL_stack_sp[-1] = sv;
-    SvREFCNT_inc_simple_void_NN(sv);
-    SvREFCNT_dec_NN(oldsv);
-    oldsv = *PL_stack_sp--;
-    SvREFCNT_dec_NN(oldsv);
+    U32 rc1 = SvREFCNT(sv1);
+    U32 rc2 = SvREFCNT(sv2);
+    /* This expression is intended to be true if either of rc1 or rc2 has
+     * the value 0 or 1, but using only a single branch test, rather
+     * than the two branches that a compiler would plant for a boolean
+     * expression. We are working on the assumption that, most of the
+     * time, neither of the args to a binary function will need to be
+     * freed - they're likely to lex vars, or PADTMPs or whatever.
+     * So give the CPU a single branch that is rarely taken. */
+    if (UNLIKELY( !(rc1>>1) + !(rc2>>1) ))
```

Comment:
I was hesitant about the obfuscation this does vs the obvious implementation, but this is pretty hot code and the Intel Optimization guide strong suggests eliminating branches where possible.

This generates 10 instructions:
```
        mov     eax, edx
        mov     edi, ecx
        shr     eax
        sete    al
        shr     edi
        movzx   eax, al
        cmp     edi, 1
        adc     eax, 0
        test    eax, eax
        jne     .L11
```

vs the straightforward version of 4:
```
        cmp     edx, 1
        jbe     .L7
        cmp     ecx, 1
        jbe     .L7

```

(this comment requires no action.)